src/com/google/enterprise/adaptor/testing/RecordingDocIdPusher.java

```diff
@@ -42,8 +43,7 @@
   private List<DocId> ids = new ArrayList<DocId>();
   private List<Record> records = new ArrayList<Record>();
   private Map<DocId, Acl> namedResources = new TreeMap<DocId, Acl>();
-  private Map<GroupPrincipal, Collection<Principal>> groups
-      = new TreeMap<GroupPrincipal, Collection<Principal>>();
+  private List<GroupPush> groupPushes = new ArrayList<GroupPush>();
```

Comment:
I was expecting a Map<String, Map<GroupPrincipal, Collection<Principal>>>, and a lightweight implementation of full and incremental pushes (we don't need to implement the swizzling, so it's just a question of replacing or adding to the inner map). This feels awkward to make assertions about. getGroupDefinitions will return incoherent mush if there are multiple group sources, and getGroupPushes is seems low-level and naval-gazing (a bit too much like mocks).

It would be nice to see this is action. We could add a RecordingDocIdPusherTest, as much to illustrate the usage as test the code (though if we do switch to a lightweight implementation, the test would be more earnest).